lib/logstash/outputs/mongodb.rb

```diff
@@ -67,22 +70,32 @@ def register
             end
           end
         end
+
+        break if @closed.true?
       end
     end
-  end # def register
+  end
 
   def receive(event)
     begin
       # Our timestamp object now has a to_bson method, using it here
       # {}.merge(other) so we don't taint the event hash innards
       document = {}.merge(event.to_hash)
+
       if !@isodate
-        # not using timestamp.to_bson
-        document["@timestamp"] = event.timestamp.to_json
+        timestamp = event.timestamp
+        if timestamp
+          # not using timestamp.to_bson
+          document["@timestamp"] = timestamp.to_json
+        else
+          @logger.warn("Cannot set MongoDB document `@timestamp` field because it does not exist in the event", :event => event)
+        end
       end
+
       if @generateId
-        document["_id"] = BSON::ObjectId.new(nil, event.timestamp)
+        document["_id"] = BSON::ObjectId.new
```

Comment:
if we _do_ have a timestamp, it should still be used as the seed for the `BSON::ObjectId`.

Reply:
- when adding the specs this `BSON::ObjectId` was actually freezing.
- looking at the `BSON::ObjectId` doc, the parameter should be an array of bytes, which probably explain the above problem.
```
Parameters:
(Array) data (defaults to: nil) — should be an array of bytes. If you want to generate a standard MongoDB object id, leave this argument blank.
```

Furthermore
```
 If no parameter is given, an id corresponding to the ObjectId BSON data type will be created. This is a 12-byte value consisting of a 4-byte timestamp, a 3-byte machine id, a 2-byte process id, and a 3-byte counter.
```

Which should be good to satisfy the intent of the option.

Reply:
I looked a bit more into the BSON code and there is a `BSON::ObjectId#from_time` method but there is a risk of creating non-unique IDs unless a `:unique => true` option is passed  and I don't really see the benefit of doing that vs a bare `BSON::ObjectId.new` which will be unique.